### PR TITLE
Async learner sanitization

### PR DESCRIPTION
**Pull request by Bam4d**
## What changes were proposed in this pull request?

1) removed the global update thread and replaced with a simpler locking mechanism
2) fixed bugs:
a) target network was not being updated unless getTargetDqnUpdate was -1 (even though it was not using DQN) 
b) target network was not getting updated after numbers of epochs, but after numbers of frames (epochs = thread total steps // nsteps) 
c) workers were being updated from the "current" global network instead of the target
3) fixed conflation between "episodes" and "epochs" an epoch counts as a single update to the global network. There can be many epochs per single episodes.
a) Epoch was being reported as episode length in logging lines.

## How was this patch tested?

A3CCartpole + AsyncNStepQLearning examples are now working and stable (large numbers of threads should not cause race conditions etc)

Unit tests are updated to pass

## Quick checklist

The following checklist helps ensure your PR is complete:

- [x] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [x] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [x] Created tests for any significant new code additions.
- [x] Relevant tests for your changes are passing.

### CODE REVIEW

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/ILearning.java
Comment (aboulang2002) [inline rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/ILearning.java:31]: We should move away from `Encodable`. Why did you add this constraint?

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncGlobal.java
Comment (aboulang2002) [inline rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncGlobal.java:95]: Thread safety: what will happens if a thread calls `getCurrent()` while another is applying gradients?
`applyGradient()` does this:
```
        valueNet.getUpdater().update(valueNet, gradient[0], valueIterationCount, valueEpochCount, batchSize, LayerWorkspaceMgr.noWorkspaces());
        valueNet.params().subi(gradient[0].gradient());
```
@AlexDBlack are `getUpdater().update()` and `params().subi()` thread safe?

If they do not throw or introduce NaNs then it'll be great; it's strong indication that Hogwild! will work perfectly in RL4J

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncGlobal.java
Comment (aboulang2002) [inline rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncGlobal.java]: `Updating target network at {} episodes`
shouldn't this be
`Updating target network at {} steps`

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThread.java
Comment (aboulang2002) [inline rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThread.java]: This is not correct. The second arg of `Math.min()` should be the number of steps remaining before we hit the max, not the actual max. From what I understand of the legacy code epoch means episode.

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThread.java
Comment (aboulang2002) [inline rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThread.java:198]: Will incorrectly return `false` if the MDP is not done but we're past the max steps per episode.

**Comment by saudet** [comment]
Please resolve the conflicts!

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThread.java
Comment (aboulang2002) [inline rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThread.java:170]: `finishEpoch()` should only be called when
`(currentEpochStep >= getConf().getMaxEpochStep() || getMdp().isDone())`

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java
Comment (aboulang2002) [inline rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java]: `nSteps * skipFrames` is not the update frequency.
`experienceSize % updateFrequency == 0` why are you doing this?

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java
Comment (aboulang2002) [inline rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java]: This is not correct. `reset()` is called when an episode starts.

**Comment by Bam4d** [comment]
> Please resolve the conflicts!

Resolved.

**Comment by saudet** [comment]
Thanks! But it looks like this PR now also contains changes from #349. Let's remove them.

**Comment by Bam4d** [comment]
> Thanks! But it looks like this PR now also contains changes from #349. Let's remove them.

The PR actually relies on the 349 changes quite heavily, we can wait until 349 is merged before finishing this review

**Comment by saudet** [comment]
Ok, but that PR doesn't build yet, so we can't merge it. Instead of creating more and more PRs, let's get those opened working! Just opening PRs and never making them actually work isn't useful.

**Review by saudet — CHANGES_REQUESTED** [review]
Please resolve the conflicts!

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/EpochStepCounter.java
Comment (saudet) [inline rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/EpochStepCounter.java:19]: Can we remove that @aboulang2002?

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncGlobal.java
Comment (saudet) [inline rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncGlobal.java]: Let's revert this to how it was.

**Review by saudet — COMMENTED** [review]
Looks good to me otherwise!

**Review by aboulang2002 — CHANGES_REQUESTED** [review]

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/ILearning.java
Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/ILearning.java:31]: Encodable is required as part of the new transformer processes you added. Its implicit that its included here currently and I wanted to make it explicit so the code is easier to understand.

Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/ILearning.java:31]: I have a further PR that updates this to push INDArray because the transformer image processing is broken at the moment

Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/ILearning.java:31]: I was going to replace it with Observer, but you said i should hold off on that

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncGlobal.java
Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncGlobal.java:95]: we shouldn't actually be copying the 'current' network in worker threads, but only grabbing the target (this is protected with a reentrant lock). The previous code was pulling the current network which it should not be doing. interence (i.e gathering experience) should only be done using the target network

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncGlobal.java
Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncGlobal.java]: Corrected this to include steps and number of worker updates (training iterations)

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThread.java
Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThread.java]: You are correct, I think i broke this during the horrible rebase i had to endure :)

Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThread.java]: Also renamed/deprecated this to getMaxStepsPerEpisode()

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThread.java
Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThread.java:198]: good catch, will add a test in for this

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThread.java
Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThread.java:170]: this is an issue of naming, finishEpoch / finishEpisode.

finishEpoch is called after every training iteration (i.e every n steps).

perhaps Epoch should be renamed trainingIteration here to remove the confusion

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java
Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java]: I agree this seems incorrect. i think a simple check of `nsteps == experience` size should be the right thing here? The previous code (before your latest merge) that did not have the experience size function needed the % to see if experience was a multiple of nsteps.

I think nsteps is also being incremented even on skipped frames which is also incorrect here. 
I will fix and add some better tests

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java
Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java]: Will change. I think this did something slightly different before you latest merge

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThread.java
Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThread.java]: have added a test to cover this case `when_remainingEpisodeLengthSmallerThanNSteps_expect_trainSubEpochCalledWithMinimumValue`

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThread.java
Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThread.java:198]: added test:`when_episodeCompletesDueToMaxStepsReached_expect_isEpisodeComplete`

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThread.java
Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThread.java:170]: As agreed the renaming of this will be on a seperate PR

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java
Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java]: added tests: `when_framesAreSkipped_expect_proportionateStepCounterUpdates` `when_episodeCompletesDueToMaxStepsReached_expect_isEpisodeComplete`
`when_episodeCompletes_expect_stepsToBeInLineWithEpisodeLenth`

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java
Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java]: Reset generateTrainingBatch also clears the training set when we update gradients with the StateActionExp Handler. Not sure we want to do this implicitly here?

Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java]: either way have added a test: `when_preEpisodeCalled_expect_experienceHandlerReset`

**Comment by saudet** [comment]
@Bam4d Please fix the copyright headers.

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/EpochStepCounter.java
Comment (aboulang2002) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/EpochStepCounter.java:19]: Yes, no problem

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java
Comment (aboulang2002) [inline rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java:95]: This should be NStep size (or the number of steps before we hit episode max steps if smaller), not the epoch/trainingSteps size.

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java
Comment (aboulang2002) [inline rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java:107]: Most n-step implementations I see are based on environment step count, not on the of gathered experience count. Why do you want to base it on experience count?

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java
Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java:95]: This is just different variable naming, the value is min(nstep, remaining episode steps).

Whats a better name here? I could just leave it as nStep, but its not always the value of nstep.

could call it `trainingIterationSteps`?

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java
Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java:107]: the number of items in the training batch is the number of step experiences we have stored. This is checking that we have collected n steps of experience. It's doing what it should do, but again, maybe the variable naming is not great.

Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java:107]: this is a much simpler way of calculating the same thing

Comment (aboulang2002) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java:107]: Edited:
~~n-step is not the count of experience collected; it's the number of environment steps taken and these two values are not equal -- if we use a random skip, your proposed change will make the loop non-deterministic.~~

I re-read the old implementation. Your version is equivalent to the old one.

Objection withdrawn.

File: rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java
Comment (aboulang2002) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java:95]: >  the value is min(nstep, remaining episode steps)
Ok that is correct. Sorry, my mistake.

What was wrong with the name `nstep`? That's what is used in litterature....

Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java:95]: I thought nstep was a bit confusing here because its really a choice between nstep and the number of remaining steps in the episode (whichever is smaller), meaning this argument is not *always* the value of nstep.

I'm happy with nStep though if you feel strongly it should be that?

Comment (aboulang2002) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java:95]: I don't want to impose my will on anything and I would normally agree with that. But I think it's important that people reading our sources be able to easily recognize the algorithms and their different parts from the literature. I think that if a term exists for a concept we should use it.

Comment (Bam4d) [reply rl4j/rl4j-core/src/main/java/org/deeplearning4j/rl4j/learning/async/AsyncThreadDiscrete.java:95]: We can talk variable names in the PR that goes through the epoch/episode/trainingStep issues

**Review by saudet — APPROVED** [review]
Looks good enough to merge for now, so let's merge this! @Bam4d We'll need to figure out something for the logs. The log entries of, for example, A3CALE with pong.bin where each episode covers multiple epochs, look funny...

**Pull request merged**
